File: ColorCoding/ColorCodingRegEx.cs
using System;
using System.Text.RegularExpressions;

namespace ColorCodingRegEx
{
    class ColorCodingRegEx
    {
        static void Main(string[] args)
        {
            int numberCouples = int.Parse(Console.ReadLine());            
            for (int n = 0; n < numberCouples; n++)
            {
                string firstRow = Console.ReadLine();
                string secondRow = Console.ReadLine();
                firstRow += " "; 
                secondRow += " ";                
                string pattern = "^" + firstRow.Replace(") ", " )?") +"$";                
                if (Regex.IsMatch(secondRow, pattern))
                {
                    Console.WriteLine("true");
                }
                else
                {
                    Console.WriteLine("false");
                }
                GC.Collect();
            }
        }
    }
}

File: ColorCoding/ColorCodingRegEx.js
function colorCoding(firstRow, secondRow) {
    firstRow += ' ';
    secondRow += ' ';
    let pattern = '^' + firstRow.replace(/\) /g, ' )?') + '$';
    let regex = new RegExp(pattern);
    if (regex.test(secondRow)) {
        return true;
    } else {
        return false;
    }
}

function colorCodingSolution(args) {
    let n = args.shift();
    for (let i = 0; i < n; i++) {
        let firstRow = args.shift().trim().replace(/\s+/g, ' ');
        let secondRow = args.shift().trim().replace(/\s+/g, ' ');
        console.log(colorCoding(firstRow, secondRow));
    }

    function colorCoding(firstRow, secondRow) {
        firstRow += ' ';
        secondRow += ' ';
        let pattern = '^' + firstRow.replace(/\) /g, ' )?') + '$';
        let regex = new RegExp(pattern);
        if (regex.test(secondRow)) {
            return true;
        } else {
            return false;
        }
    }
}

colorCodingSolution([
    1,
    '(purple) (green) (purple) red blue',
    'green red blue'
]);

colorCodingSolution([
    2,
    '(blue) green (red) (blue)',
    'blue red blue',
    '(purple) red (blue) (yellow) blue (purple) (green) black',
    'red blue green black'
]);
